table/tables/partition.go

```diff
@@ -260,8 +260,17 @@ func (t *partitionedTable) locateRangePartition(ctx sessionctx.Context, pi *mode
 		idx = 0
 	}
 	if idx < 0 || idx >= len(partitionExprs) {
-		// The data does not belong to any of the partition?
-		return 0, errors.Trace(table.ErrTrgInvalidCreationCtx)
+		// The data does not belong to any of the partition returns `table has no partition for value %s`.
+		e, err := expression.ParseSimpleExprWithTableInfo(ctx, pi.Expr, t.meta)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		
+		ret, _, err2 := e.EvalInt(ctx, chunk.MutRowFromDatums(r).ToRow())
+		if err2 != nil {
+			return 0, errors.Trace(err2)
+		}
+		return 0, errors.Trace(table.ErrNoPartitionForGivenValue.GenWithStackByArgs(fmt.Sprintf("%d", ret)))
```

Comment:
`table has no partition for value %s`

The value is the partition id rather than the original value ?

Reply:
Is the original value.

Reply:
@tiancaiamao 
```sql
mysql>  CREATE TABLE shard_table_tr (id INT, name VARCHAR(50), purchased DATE)
    ->     PARTITION BY RANGE( YEAR(purchased) ) (
    ->     PARTITION p0 VALUES LESS THAN (1990),
    ->     PARTITION p1 VALUES LESS THAN (1995),
    ->     PARTITION p2 VALUES LESS THAN (2000),
    ->     PARTITION p3 VALUES LESS THAN (2005),
    ->     PARTITION p4 VALUES LESS THAN (2010),
    ->     PARTITION p5 VALUES LESS THAN (2015)
    ->    );
Query OK, 0 rows affected (0.01 sec)

mysql>  INSERT INTO shard_table_tr VALUES
    ->    (1, 'desk organiser', '2003-10-15'),
    ->    (2, 'alarm clock', '1997-11-05'),
    ->    (3, 'chair', '2009-03-10'),
    ->    (4, 'bookcase', '1989-01-10'),
    ->     (5, 'exercise bike', '2014-05-09'),
    ->     (6, 'sofa', '1987-06-05'),
    ->     (7, 'espresso maker', '2011-11-22'),
    ->     (8, 'aquarium', '1992-08-04'),
    ->     (9, 'study desk', '2006-09-16'),
    ->     (10, 'lava lamp', '1998-12-25');
Query OK, 10 rows affected (0.00 sec)
Records: 10  Duplicates: 0  Warnings: 0

mysql> INSERT INTO shard_table_tr VALUES (10, 'lava lamp', '2998-12-25');
ERROR 1526 (HY000): Table has no partition for value 2998
```